File: api/src/main/java/xtdb/util/NormalForm.java
package xtdb.util;

import clojure.lang.Keyword;
import clojure.lang.Symbol;

import java.util.Locale;

public class NormalForm {
    private NormalForm() {
    }

    public static String snakeCase(String s) {
        var i = s.lastIndexOf('$');
        if (i < 0) {
            return s.toLowerCase(Locale.ROOT)
                    .replace("$", "/")
                    .replace("-", "_");
        } else {
            return String.format("%s/%s",
                    s.substring(0, i).replace("$", ".").replace("-", "_"),
                    s.substring(i + 1).replace("-", "_"));
        }
    }

    private static String normalise(String s) {
        return s.toLowerCase(Locale.ROOT)
                .replace('.', '$')
                .replace('-', '_');
    }

    public static String normalForm(String s) {
        var i = s.lastIndexOf('/');
        if (i < 0) {
            return normalise(s);
        } else {
            return String.format("%s$%s", normalise(s.substring(0, i)), normalise(s.substring(i + 1)));
        }
    }

    public static Symbol normalForm(Symbol sym) {
        if (sym.getNamespace() != null) {
            return Symbol.intern(String.format("%s$%s", normalise(sym.getNamespace()), normalise(sym.getName())));
        } else {
            return Symbol.intern(normalise(sym.getName()));
        }
    }

    public static Keyword normalForm(Keyword k) {
        return Keyword.intern(normalForm(k.sym));
    }

}

File: api/src/main/kotlin/xtdb/util/NormalForm.kt
@file:JvmName("NormalForm")

package xtdb.util

import clojure.lang.Keyword
import clojure.lang.Symbol
import com.github.benmanes.caffeine.cache.Caffeine

internal fun normalForm0(s: String): String = s
    .replace('-', '_')
    .replace(Regex("^xt/"), "_")
    .split('.', '/', '$')
    .joinToString(separator = "$")
    .lowercase()

private val NORMAL_FORM_CACHE = Caffeine.newBuilder()
    .maximumSize(1 shl 15)
    .build<String, String>(::normalForm0)

/**
 * @suppress
 */
fun normalForm(s: String): String = NORMAL_FORM_CACHE[s]

/**
 * @suppress
 */
fun normalForm(sym: Symbol): Symbol =
    Symbol.intern(
        when (sym.namespace) {
            null -> normalForm(sym.name)
            "xt" -> "_${normalForm(sym.name)}"
            else -> "${normalForm(sym.namespace)}$${normalForm(sym.name)}"
        }
    )

/**
 * @suppress
 */
@Suppress("unused") // Clojure
fun normalForm(k: Keyword): Keyword = Keyword.intern(normalForm(k.sym))

private val String.normalTableSegment get() = lowercase().replace('-', '_')

fun Keyword.normalTableName(): Symbol =
    Symbol.intern(namespace?.normalTableSegment ?: "public", name.normalTableSegment)

/**
 * @suppress
 */
internal fun String.kebabToCamelCase(): String {
    val pattern = "-[a-z]".toRegex()
    return replace(pattern) { it.value.last().uppercase() }
}
